S3Request.gs

```diff
@@ -135,128 +157,259 @@ S3Request.prototype.execute = function(options) {
+}
+
+S3Request.prototype.stringToSign = function(datetime) {
+  var parts = [];
+  parts.push('AWS4-HMAC-SHA256');
+  parts.push(datetime);
+  parts.push(this.credentialString(datetime));
+  parts.push(this.hexEncodedHash(this.canonicalString()));
+  return parts.join('\n');
+}
+
+S3Request.prototype.canonicalString = function() {
+  var parts = [];
+  var [base, search] = this.getUrl().split("?", 2)
+  parts.push(this.httpMethod);
+  parts.push(this.canonicalUri(base));
+  parts.push(this.canonicalQueryString(search));
+  parts.push(this.canonicalHeaders() + '\n');
+  parts.push(this.signedHeaders());
+  parts.push(this.hexEncodedBodyHash());
+  return parts.join('\n');
+}
+
+S3Request.prototype.canonicalUri = function(uri) {
+  var m = uri.match(/https?:\/\/(.+)\.s3.*\.amazonaws\.com\/(.+)$/);
+  var object = m ? m[2] : ""
+  return "/" + encodeURIComponent(object).replace(/%2F/ig, '/')
+}
+
+S3Request.prototype.canonicalQueryString = function(values) {
+  if (!values) return ""
+  var parts = [];
+  var items = values.split("&");
+  for (var i in items) {
+    var [key, value] = items[i].split("=")
+    parts.push(encodeURIComponent(key.toLowerCase()) + "=" + encodeURIComponent(value))
+  }
+  return parts.sort().join("&")
+}
+
+S3Request.prototype.canonicalHeaders = function() {
+  var parts = [];
+  for (var item in this.headers) {
+    var key = item.toLowerCase();
+    if (this.isSignableHeader(key)) {
+      var header = key + ":" + this.canonicalHeaderValues(this.headers[item].toString())
+      parts.push(header)
+    }
+  }
+  return parts.sort().join("\n")
+}
+
+S3Request.prototype.canonicalHeaderValues = function(values) {
+  return values.replace(/\s+/g, " ").trim();
+}
+
+S3Request.prototype.signedHeaders = function() {
+  var keys = [];
+  for (var key in this.headers) {
+    key = key.toLowerCase();
+    if (this.isSignableHeader(key)) {
+      keys.push(key);
+    }
+  }
+  return keys.sort().join(';');
+}
+
+S3Request.prototype.credentialString = function(datetime) {
+  return [
+    datetime.substr(0, 8),
+    this.region,
+    this.serviceName,
+    'aws4_request'
+  ].join('/');
+}
+
+S3Request.prototype.hexEncodedHash = function(string) {
+  return this.hex(Utilities.computeDigest(Utilities.DigestAlgorithm.SHA_256, string, Utilities.Charset.UTF_8));
+}
+
+S3Request.prototype.hexEncodedBodyHash = function() {
+  if (this.isPresigned() && !this.content.length) {
+    return 'UNSIGNED-PAYLOAD'
+  } else if (this.headers['X-Amz-Content-Sha256']) {
+    return this.headers['X-Amz-Content-Sha256']
   } else {
-    return ""; 
+    return this.hexEncodedHash(this.content || '')
   }
-};
+}
+
+S3Request.prototype.isSignableHeader = function(key) {
+  if (key.match(/^x-amz/i)) return true;
+  var unsignableHeaders = [
+    'authorization',
+    'content-type',
+    'content-length',
+    'user-agent',
+    this.expiresHeader,
+    'expect',
+    'x-amzn-trace-id'
+  ];
+  return unsignableHeaders.indexOf(key) < 0
```

Comment:
i think you need to lower-case `key` to test this correctly. HTTP headers are case insensitive. Also I think comparison would be better testing the whole string for equivalence, rather than whether header name begins with name of a header that you know to be unsignable.

Reply:
I agree. I fix.